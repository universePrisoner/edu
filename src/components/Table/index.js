import React, { Component } from "react";
import PropTypes from 'prop-types';
import { sortBy } from 'lodash';
import classNames from 'classnames';
import "./index.css";

/**
 * Import component
 */
import { Button } from '../Button';

const SORT = {
    NONE: list => list,
    TITLE: list => sortBy(list, 'title'),
    AUTHOR: list => sortBy(list, 'author'),
    COMMENTS: list => sortBy(list, 'num_comments').reverse(),
    POINTS: list => sortBy(list, 'points').reverse()
}

const Sort = ({ sortKey, activeSortKey, onSort, children }) => {
    const sortClass = classNames(
        'button-inline',
        { 'button-active': sortKey === activeSortKey }
    );

    return (<Button onClick={() => onSort(sortKey)} className={sortClass}>{children}</Button>);
}

class Table extends Component {
    constructor(props) {
        super(props);

        this.state = {};
    }

    render() {
        const {
            list,
            sortKey,
            isSortReverse,
            onSort,
            onDismiss
        } = this.props;

        const sortedList = SORT[sortKey](list);
        const reverseSortedList = isSortReverse ? sortedList.reverse() : sortedList;

        return (
            <div className="table">
                <div className="table-header">
                    <span className="large-column">
                        <Sort sortKey={'TITLE'} onSort={onSort} activeSortKey={sortKey}>Title</Sort>
                    </span>
                    <span className="middle-column">
                        <Sort sortKey={'AUTHOR'} onSort={onSort} activeSortKey={sortKey}>Author</Sort>
                    </span>
                    <span className="small-column">
                        <Sort sortKey={'COMMENTS'} onSort={onSort} activeSortKey={sortKey}>Comments</Sort>
                    </span>
                    <span className="small-column">
                        <Sort sortKey={'POINTS'} onSort={onSort} activeSortKey={sortKey}>Points</Sort>
                    </span>
                </div>

                {
                    reverseSortedList.map(item =>
                        <div key={item.objectID} className="table-row">
                            <span className="large-column">
                                <a href={item.url} >{item.title}</a>
                            </span>
                            <span className="middle-column">{item.author}</span>
                            <span className="small-column">{item.num_comments}</span>
                            <span className="small-column">{item.points}</span>
                            <span className="small-column">
                                <Button onClick={() => onDismiss(item.objectID)} className="button-inline">Drop It</Button>
                            </span>
                        </div>
                    )
                }
            </div>);
    }
};

Table.propTypes = {
    list: PropTypes
        .arrayOf(
            PropTypes.shape({
                objectID: PropTypes.string.isRequired,
                author: PropTypes.string,
                url: PropTypes.string,
                num_comments: PropTypes.number,
                points: PropTypes.number,
            })
        )
        .isRequired,
    onDismiss: PropTypes.func.isRequired
}

export default Table;