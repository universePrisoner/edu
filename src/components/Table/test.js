import React from 'react';
import ReactDOM from 'react-dom';
import renderer from "react-test-renderer";
import Enzyme, { shallow } from "enzyme";
import Adapter from "enzyme-adapter-react-16";
import Table from './index';

Enzyme.configure({ adapter: new Adapter() });

describe('Table', () => {
    const onDismiss = () => console.log(1);

    const props = {
        list: [
            { title: '1', author: '1', num_comments: 1, points: 2, objectID: 'y' },
            { title: '2', author: '2', num_comments: 2, points: 4, objectID: 'x' },
        ],
        sortKey: 'AUTHOR',
        onSort: () => console.log(1),
        onDismiss: () => console.log(1),
        isSortReverse: false
    }

    it('renders without crashing ', () => {
        const div = document.createElement('div');
        ReactDOM.render(<Table {...props} onDismiss={onDismiss} />, div);
        ReactDOM.unmountComponentAtNode(div);
    });

    test('table short snapshot', () => {
        const component = renderer.create(<Table {...props} onDismiss={onDismiss} />);
        const tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('shows two items in list', () => {
        const element = shallow(<Table {...props} onDismiss={onDismiss} />)

        expect(element.find('.table-row').length).toBe(2);
    });
});