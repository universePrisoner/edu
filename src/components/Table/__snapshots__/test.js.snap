// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table table short snapshot 1`] = `
<div
  className="table"
>
  <div
    className="table-header"
  >
    <span
      className="large-column"
    >
      <button
        className="button-inline"
        onClick={[Function]}
        type="button"
      >
        Title
      </button>
    </span>
    <span
      className="middle-column"
    >
      <button
        className="button-inline button-active"
        onClick={[Function]}
        type="button"
      >
        Author
      </button>
    </span>
    <span
      className="small-column"
    >
      <button
        className="button-inline"
        onClick={[Function]}
        type="button"
      >
        Comments
      </button>
    </span>
    <span
      className="small-column"
    >
      <button
        className="button-inline"
        onClick={[Function]}
        type="button"
      >
        Points
      </button>
    </span>
  </div>
  <div
    className="table-row"
  >
    <span
      className="large-column"
    >
      <a>
        1
      </a>
    </span>
    <span
      className="middle-column"
    >
      1
    </span>
    <span
      className="small-column"
    >
      1
    </span>
    <span
      className="small-column"
    >
      2
    </span>
    <span
      className="small-column"
    >
      <button
        className="button-inline"
        onClick={[Function]}
        type="button"
      >
        Drop It
      </button>
    </span>
  </div>
  <div
    className="table-row"
  >
    <span
      className="large-column"
    >
      <a>
        2
      </a>
    </span>
    <span
      className="middle-column"
    >
      2
    </span>
    <span
      className="small-column"
    >
      2
    </span>
    <span
      className="small-column"
    >
      4
    </span>
    <span
      className="small-column"
    >
      <button
        className="button-inline"
        onClick={[Function]}
        type="button"
      >
        Drop It
      </button>
    </span>
  </div>
</div>
`;
